blocksync/buffer.go

```diff
@@ -30,89 +30,57 @@ type blockBuffer struct {
 func (b *blockBuffer) Flush(blk *blockchain.Block) (bool, bCheckinResult) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	l := logger.With().Uint64("recvHeight", blk.Height()).Uint64("startHeight", b.startHeight).Uint64("confirmedHeight", b.confirmedHeight).Str("source", "blockBuffer").Logger()
-
-	var (
-		syncedHeight uint64
-		moved        bool
-	)
-
+	moved := false
 	// check
 	h := blk.Height()
 	if h <= b.confirmedHeight {
 		return moved, bCheckinLower
 	}
-	if h < b.startHeight {
-		if err := commitBlock(b.bc, b.ap, blk); err != nil {
-			return moved, bCheckinLower
-		}
-
-		b.confirmedHeight = h
-		b.startHeight = h + 1
-		l.Info().Uint64("confirmedHeight", b.confirmedHeight).Msg("replace old dummy.")
-		return true, bCheckinValid
-	}
-	if b.blocks[h] != nil {
+	if blk, ok := b.blocks[h]; ok && blk != nil {
 		return moved, bCheckinExisting
 	}
-	if b.startHeight+b.size <= h {
+	if h > b.confirmedHeight+b.size {
 		return moved, bCheckinHigher
 	}
 	b.blocks[h] = blk
-
-	for syncHeight := b.startHeight; syncHeight < b.startHeight+b.size; syncHeight++ {
-		if b.blocks[syncHeight] == nil {
-			continue
-		}
-		if err := commitBlock(b.bc, b.ap, b.blocks[syncHeight]); err == nil {
-			syncedHeight = syncHeight
-			b.confirmedHeight = syncedHeight
-			l.Info().Uint64("syncedHeight", syncedHeight).Msg("Successfully committed block.")
-			delete(b.blocks, syncHeight)
-			continue
-		} else {
-			l.Error().Err(err).Msg("Failed to commit the block")
+	l := logger.With().
+		Uint64("recvHeight", blk.Height()).
+		Uint64("confirmedHeight", b.confirmedHeight).
+		Str("source", "blockBuffer").
+		Logger()
+	for b.size > 0 {
+		next := b.confirmedHeight + 1
+		blk, ok := b.blocks[next]
+		if !ok || blk == nil {
+			break
 		}
-
-		// unable to commit, check reason
-		if blk, err := b.bc.GetBlockByHeight(syncHeight); err == nil {
-			if blk.HashBlock() == b.blocks[syncHeight].HashBlock() {
-				// same existing block
-				syncedHeight = syncHeight
-				b.confirmedHeight = syncedHeight
-			}
-			delete(b.blocks, syncHeight)
-		} else {
-			th := b.bc.TipHeight()
-			if syncHeight == th+1 {
-				// bad block or forked here
-				l.Error().Uint64("syncHeight", syncHeight).
-					Uint64("syncedHeight", syncedHeight).
-					Uint64("tipHeight", th).
-					Msg("Failed to commit next block.")
-				delete(b.blocks, syncHeight)
+		delete(b.blocks, next)
+		if err := commitBlock(b.bc, b.ap, blk); err != nil {
+			l.Error().Err(err).Uint64("syncHeight", next).
+				Msg("Failed to commit the block.")
+			// unable to commit, check reason
+			committedBlk, err := b.bc.GetBlockByHeight(next)
+			if err != nil || committedBlk.HashBlock() != blk.HashBlock() {
+				break
 			}
-			// otherwise block is higher than currently height
 		}
-	}
-	if syncedHeight != 0 {
-		b.startHeight = syncedHeight + 1
 		moved = true
+		b.confirmedHeight = next
+		l.Info().Uint64("syncedHeight", next).Msg("Successfully committed block.")
 	}
 
 	// clean up on memory leak
```

Comment:
i am wondering we actually don't need this check? since blocks < tip and > tip + b.size are discarded in Flush(), map can only contain blocks from tip to tip + b.size